components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/CertificateBasedTokenBinder.java

```
@@ -0,0 +1,212 @@
+
+        String cnfValue = generateCnfHashValue((HttpServletRequest) request);
+        if (StringUtils.isNotBlank(cnfValue)) {
+            return bindingReference.equals(OAuth2Util.getTokenBindingReference(cnfValue));
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isValidTokenBinding(OAuth2AccessTokenReqDTO oAuth2AccessTokenReqDTO, String bindingReference) {
+
+        String refreshToken = oAuth2AccessTokenReqDTO.getRefreshToken();
+        try {
+            TokenBinding tokenBinding = getBindingFromRefreshToken(refreshToken, OAuth2Util.isHashEnabled());
+            String cnfValue = generateCnfHashValue(oAuth2AccessTokenReqDTO.getHttpServletRequestWrapper());
+
+            if (tokenBinding != null && CERTIFICATE_BASED_TOKEN_BINDER.equals(tokenBinding.getBindingType()) &&
+                    StringUtils.isNotBlank(cnfValue)) {
+                return cnfValue.equalsIgnoreCase(tokenBinding.getBindingValue()) && bindingReference
+                        .equalsIgnoreCase(tokenBinding.getBindingReference());
+            }
+            return false;
+        } catch (IdentityOAuth2Exception e) {
+            return false;
+        }
+    }
+
+    private String generateCnfHashValue(HttpServletRequest request) {
+
+        Base64URL certThumbprint;
+        X509Certificate certificate = null;
+        String headerName = Optional.ofNullable(IdentityUtil.getProperty(OAuthConstants.MTLS_AUTH_HEADER))
+                .orElse(OAuthConstants.CONFIG_NOT_FOUND);
+
+        String certificateInHeader = request.getHeader(headerName);
+        Object certObject = Optional.ofNullable(request.getAttribute(OAuthConstants.JAVAX_SERVLET_REQUEST_CERTIFICATE))
+                .orElse(null);
+
+        if (StringUtils.isNotBlank(certificateInHeader)) {
+            try {
+                certificate = parseCertificate(certificateInHeader);
+            } catch (CertificateException e) {
+                return null;
```

Better to throw the exception and handle it